staging/src/k8s.io/apiserver/pkg/server/openapi/openapi.go

```diff
@@ -17,35 +17,68 @@ limitations under the License.
 package openapi
 
 import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
+	"gopkg.in/yaml.v2"
+	"mime"
 	"net/http"
 	"reflect"
 	"strings"
+	"time"
 
 	"github.com/emicklei/go-restful"
 	"github.com/go-openapi/spec"
+	"github.com/golang/protobuf/proto"
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
 
 	"k8s.io/apimachinery/pkg/openapi"
-	"k8s.io/apimachinery/pkg/util/json"
 	genericmux "k8s.io/apiserver/pkg/server/mux"
 	"k8s.io/apiserver/pkg/util/trie"
 )
 
 const (
 	OpenAPIVersion  = "2.0"
 	extensionPrefix = "x-kubernetes-"
+
+	JSON_EXT = ".json"
+
+	MIME_JSON  = "application/json"
+	MIME_PB    = "application/com.github.googleapis.gnostic.OpenAPIv2+protobuf"
+	MIME_PB_GZ = "application/x-gzip"
 )
 
 type openAPI struct {
 	config       *openapi.Config
 	swagger      *spec.Swagger
+	swaggerBytes []byte
+	swaggerPb    []byte
+	swaggerPbGz  []byte
+	lastModified time.Time
 	protocolList []string
 	servePath    string
 	definitions  map[string]openapi.OpenAPIDefinition
 }
 
+func computeEtag(data []byte) string {
+	sha := sha512.Sum512(data)
+	return fmt.Sprintf("\"%s\"",
```

Comment:
This looks like a good situation to either use `%q` or a backtick string literal.